lib/internal/modules/esm/module_job.js

```diff
@@ -111,6 +89,38 @@ class ModuleJob {
     this.instantiated = undefined;
   }
 
+  /**
+   * Iterates the module requests and links with the loader.
+   * @returns {Promise<ModuleJob[]>} Dependency module jobs.
+   */
+  async _link() {
+    this.module = await this.modulePromise;
+    assert(this.module instanceof ModuleWrap);
+
+    // Explicitly keeping track of dependency jobs is needed in order
+    // to flatten out the dependency graph below in `_instantiate()`,
+    // so that circular dependencies can't cause a deadlock by two of
+    // these `link` callbacks depending on each other.
+    const dependencyJobs = [];
+    const promises = [];
+    // Iterate with index to avoid calling into userspace with `Symbol.iterator`.
+    const moduleRequestsLength = this.module.moduleRequests.length;
+    for (let idx = 0; idx < moduleRequestsLength; idx++) {
+      const { specifier, attributes } = this.module.moduleRequests[idx];
+      const job = await this.loader.getModuleJob(specifier, this.url, attributes);
```

Comment:
changing this to a `for` loop makes each `getModuleJob` run in series. also this code is still jumping between js<->c++ for every module.
what if you did something like
```js
const dependencyJobs = [];
const promises = this.module.moduleRequests.map(async ({specifier, attributes}) => {
  const job = await this.loader.getModuleJob(specifier, this.url, attributes);
  dependencyJobs.push(job);
  return job.modulePromise;
});
this.module.link(promises); // pass array and let c++ iterate it
await Promise.all(promises);
return Promise.all(dependencyJobs);
```

Reply:
Thanks for spotting this! Fixed with a test added.